sql/catalyst/src/main/scala/org/apache/spark/sql/sources/filters.scala

```diff
@@ -32,6 +33,7 @@ import org.apache.spark.annotation.{Evolving, Stable}
 sealed abstract class Filter {
   /**
    * List of columns that are referenced by this filter.
+   * Note that, if a column contains `dots` in name, it will be quoted to avoid confusion.
```

Comment:
https://github.com/apache/spark/pull/27728/files#r390853911, I think it shouldn't be a legacy configuration but a proper configuration might list which source will take the nested filter-push down.

There is no workaround for the behaviour change except this legacy configuration; but legacy configurations are supposed to be removed. If we're going to add this as a legacy configuration, we should have a way to don't unquote this.

Quoting itself might not exist in some downstream datasource implementations. Dots can be used for different meanings such as namespaces. Some source don't have nested structures at all and presumably they won't also have such quotes at all.

Reply:
I agree that some downstream datasoruce implementation might not support column name containing `dots` or nested column predicate pushdown. However, we are talking about the contract of Spark's Filter API. The implementation of the translation logics for each datasources should take care of this properly instead, and that's why we have this abstraction. WDYT?

Reply:
That's logically correct if we introduce the very first version of this API. I think I more mean the backward compatibility.

There are many DSv1 implementations arguably ~90% vs DSv2 up to my best knowledge, and I worry if people start to complain it doesn't work and there's no workaround but their implementations should unquote by themselves specifically in Spark 3.

Reply:
As @cloud-fan suggested, I add a conf to disable the new behavior so people can still use their old DSv1 implementation while upgrading to the new behavior.

Reply:
I think the problem is that, it is one switch for all or none while the problem can exist for source specifically.

I feel like this concern is underestimated. Dots in column names are not very common in Spark world but I don't think it's not very rare in other systems, Elasticsearchm as an example (https://github.com/elastic/elasticsearch/issues/15951). 

Another concern is that, it's enabled by default. If we target 3.0, should we maybe disable it by default? It's already right before cutting RC very soon.